docs/tutorials/chatbot.md

```diff
@@ -0,0 +1,728 @@
+        fill_val = torch.tensor(
+            fill_vals[self.pooling_strategy], device=torch.device('cpu')
+        )
+
+        layer = hidden_states[self.layer_index]
+        attn_mask = input_tokens['attention_mask'].unsqueeze(-1).expand_as(layer)
+        layer = torch.where(attn_mask.bool(), layer, fill_val)
+
+        embeddings = layer.sum(dim=1) / attn_mask.sum(dim=1)
+        return embeddings.cpu().numpy()
+
+    @requests
+    def encode(self, docs: 'DocumentArray', *args, **kwargs):
+        import torch
+
+        with torch.no_grad():
+
+            if not self.tokenizer.pad_token:
+                self.tokenizer.add_special_tokens({'pad_token': '[PAD]'})
+                self.model.resize_token_embeddings(len(self.tokenizer.vocab))
+
+            input_tokens = self.tokenizer(
+                docs.get_attributes('content'),
+                max_length=self.max_length,
+                padding='longest',
+                truncation=True,
+                return_tensors='pt',
+            )
+            input_tokens = {
+                k: v.to(torch.device('cpu')) for k, v in input_tokens.items()
+            }
+
+            outputs = getattr(self.model, self.embedding_fn_name)(**input_tokens)
+            if isinstance(outputs, torch.Tensor):
+                return outputs.cpu().numpy()
+            hidden_states = outputs.hidden_states
+
+            embeds = self._compute_embedding(hidden_states, input_tokens)
+            for doc, embed in zip(docs, embeds):
+                doc.embedding = embed
+```
+
+`MyTransformer` exposes only one endpoint: `encode`. This will be called whenever we make a request to the flow, either 
+on query or index. The endpoint will create embeddings for the indexed or query documents so that they can be used 
+to get the closed matches.
+```{admonition} Note
+:class: note
+Encoding is a fundamental concept in neural search. It means representing the data in a vectorial form (embeddings).
+```
+
+Encoding is performed through a transformers model (`sentence-transformers/paraphrase-mpnet-base-v2` by default).
+
+Then, we can implement our indexer (`MyIndexer`):
+```python
+class MyIndexer(Executor):
+    """Simple indexer class """
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self._docs = DocumentArrayMemmap(self.workspace + '/indexer')
+
+    @requests(on='/index')
+    def index(self, docs: 'DocumentArray', **kwargs):
+        self._docs.extend(docs)
+
+    @requests(on='/search')
+    def search(self, docs: 'DocumentArray', **kwargs):
+        """Append best matches to each document in docs
+
+        :param docs: documents that are searched
+        :param parameters: dictionary of pairs (parameter,value)
+        :param kwargs: other keyword arguments
+        """
+        docs.match(
+            self._docs,
+            metric='cosine',
+            normalization=(1, 0),
+            limit=1,
```

Comment:
This can comfortably fit all on one line

Reply:
done